odoo/models.py

```diff
@@ -3967,6 +3989,95 @@ def _write(self, vals):
 
         return True
 
+    def _add_missing_computes(self, vals, stored_computed_fields, x2m_stored_fields):
+        """ Adds missing computed fields to data_list values.
+        Only applies for pre_compute=True fields.
+        :param dict vals: dict of create values
+        :return: updated vals, with pre_computed fields of current model
+            And pre_computed fields of x2m models
+        """
+
+        if not stored_computed_fields:
+            return
+
+        missing_fields = [
+            field for field in stored_computed_fields
+            if field.name not in vals
+        ]
+
+        if missing_fields:
+            # Specific context to know when we are in this specific case.
+            new_obj = self.with_context(create_pre_compute=True).new(vals)
+
+            # While we already have a new of the main record,
+            # Try to pre compute the computed fields of its x2m fields.
+            x2m_precomputables = [
+                field for field in x2m_stored_fields
+                if field.name in vals
+            ]
+
+            for field in x2m_precomputables:
+                # Trigger pre_computation of x2m pre_compute=True fields
+                # Allow batch computation for models with lots of x2m content
+                # SO with K lines, all lines pre_compute are computed in the new
+                # SO, avoiding another new for all lines.
+                updated = new_obj[field.name]._rec_force_precomputes(field)
+                if updated:
+                    # If the x2m records are supposed updated, keep their
+                    # values for the create to avoid useless recomputations.
+                    field_value = field.convert_to_write(new_obj[field.name], self)
+                    vals[field.name] = field_value
+
+            for field in missing_fields:
+                # computed stored fields with a column
+                # have to be computed before create
+                # s.t. required and constraints can be applied on those fields.
+                field_value = field.convert_to_write(new_obj[field.name], self)
+                vals[field.name] = field_value
+
+        return vals
+
+    def _rec_force_precomputes(self, source_field):
+        """Recursively force precomputation of x2m fields pre_computed fields.
```

Comment:
`_rec` may be too ambiguous as a prefix for a function, overall I don't think the fact that it is recursive should be a part of the function's name..